x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/index/query/PinnedQueryBuilder.java

```diff
@@ -29,34 +32,95 @@
     public static final String NAME = "pinned";
     protected final QueryBuilder organicQuery;
     protected final List<String> ids;
+    private final List<Item> docs;
     protected static final ParseField IDS_FIELD = new ParseField("ids");
+    private static final ParseField DOCS_FIELD = new ParseField("docs");
     protected static final ParseField ORGANIC_QUERY_FIELD = new ParseField("organic");
 
+    /**
+     * A single item to be used for a {@link PinnedQueryBuilder}.
+     */
+    public static final class Item implements ToXContentObject, Writeable {
+        private static final ParseField INDEX_FIELD = new ParseField("_index");
+        private static final ParseField ID_FIELD = new ParseField("_id");
+
+        private final String index;
+        private final String id;
+
+        public Item(String index, String id) {
+            if (index == null) {
+                throw new IllegalArgumentException("Item requires index to be non-null");
+            }
+            if (Regex.isSimpleMatchPattern(index)) {
+                throw new IllegalArgumentException("Item index cannot contain wildcard expressions");
+            }
+            if (id == null) {
+                throw new IllegalArgumentException("Item requires id to be non-null");
+            }
+            this.index = index;
+            this.id = id;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(index);
+            out.writeString(id);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field(INDEX_FIELD.getPreferredName(), this.index);
```

Comment:
Small comment, we usually don't use the `this` reference unless it helps avoid ambiguity (like in a constructor).

Reply:
That makes sense. It wasn't clear to me when `this` was necessary. I removed it here but left it in other spots where it was already used.